x/concentrated-liquidity/swaps.go

```diff
@@ -808,3 +808,122 @@ func edgeCaseInequalityBasedOnSwapStrategy(isZeroForOne bool, nextInitializedTic
 	}
 	return nextInitializedTickSqrtPrice.LT(computedSqrtPrice)
 }
+
+// ComputeMaxInAmtGivenMaxTicksCrossed calculates the maximum amount of the tokenInDenom that can be swapped
+// into the pool to swap through all the liquidity from the current tick through the maxTicksCrossed tick,
+// but not exceed it.
+func (k Keeper) ComputeMaxInAmtGivenMaxTicksCrossed(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenInDenom string,
+	maxTicksCrossed uint64,
+) (maxTokenIn, resultingTokenOut sdk.Coin, err error) {
+	cacheCtx, _ := ctx.CacheContext()
+
+	p, err := k.getPoolForSwap(cacheCtx, poolId)
+	if err != nil {
+		return sdk.Coin{}, sdk.Coin{}, err
+	}
+
+	// Validate tokenInDenom exists in the pool
+	if tokenInDenom != p.GetToken0() && tokenInDenom != p.GetToken1() {
+		return sdk.Coin{}, sdk.Coin{}, types.TokenInDenomNotInPoolError{TokenInDenom: tokenInDenom}
+	}
+
+	// Determine the tokenOutDenom based on the tokenInDenom
+	var tokenOutDenom string
+	if tokenInDenom == p.GetToken0() {
+		tokenOutDenom = p.GetToken1()
+	} else {
+		tokenOutDenom = p.GetToken0()
+	}
+
+	// Setup the swap strategy
+	swapStrategy, _, err := k.setupSwapStrategy(p, p.GetSpreadFactor(cacheCtx), tokenInDenom, sdk.ZeroDec())
+	if err != nil {
+		return sdk.Coin{}, sdk.Coin{}, err
+	}
+
+	// Initialize swap state
+	swapState := newSwapState(types.MaxSpotPrice.RoundInt(), p, swapStrategy)
```

Comment:
Why do we use types.MaxSpotPrice? Do we decrement it until maxTicksCrossed ?

Reply:
Yeah we basically never want the reason we stop an estimation to be due to a lack of funds, it should only ever strictly be due to the limit imposed by amt of ticks willing to traverse.

Reply:
Yeah I think that makes sense too, I can change it to that